sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -495,6 +495,25 @@ class Dataset[T] private[sql](
     select(newCols : _*)
   }
 
+  /**
+   * Returns a new Dataset where each record has been mapped on to the specified type.
+   * This only supports `U` being a class. Fields for the class will be mapped to columns of the
+   * same name (case sensitivity is determined by `spark.sql.caseSensitive`).
+   *
+   * If the schema of the Dataset does not match the desired `U` type, you can use `select`
+   * along with `alias` or `as` to rearrange or rename as required.
+   *
+   * This method eagerly projects away any columns that are not present in the specified class.
+   * It further guarantees the order of columns as well as data types to match `U`.
+   *
+   * @group basic
+   * @since 3.0.0
+   */
+  def toDS[U : Encoder]: Dataset[U] = {
+    val columns = implicitly[Encoder[U]].schema.fields.map(f => col(f.name).cast(f.dataType))
```

Comment:
what about inner fields? e.g.
```
case class Inner(a: Int)
case class Outer(i: Inner)
```
and the schema is `<col: struct<a: int, b: int>>`

Reply:
and the encoder can do anything, e.g. `Encoders.javaSerialization`, whose schema is just a binary column.

I think the only way is to execute the encoder and do a round trip, which is effectively `df.as[T].map(identity)`

Reply:
As the documentation of that method says, this only supports U being a class. Tuples or structs are not supported. Those would require knowledge that only the encoder does have.

For classes a simple projection is sufficient, which is very cheap. An encoder round trip should be avoided in this use case. The stricter as method could be extended later to other encoders and kinds of Us, but I think classes are a wide use case that justify adding it already.

Reply:
This doesn't work even for simple classes when the encoder is `Encoders.javaSerialization`. I agree that this is a valid use case, and people can create a Util function to do it. But I don't agree this is general enough to be put in a method named `toDS`.

Reply:
Do you agree that the lazy-ness of `as[T]` is an inconsistency of the API where all other transformations' schema instantly reflect the result's schema? With `as[T]` you need a `map(identity)` to really become a `Dataset[T]`.

Which properties of `toDS[T]` would you require?
# case classes, tuples and inner fields supported
# arbitrary implementations of `Encoder` supported
# data (not just schema) of result `Dataset[T]` are identical to `as[T].map(identity)`, which is given when they follow schema `encoder.schema`
# uses projection where possible (efficient), falls back to `map(identity)` when needed (expensive)

The idea is to turn the dataset into the same shape as it would look like when the encoder would produce it (`map(identity)`). In case of `ExpressionEncoder`, the encoder can tell you which columns it serialises / deserialises. If the schema of those columns does not change (e.g. unchanged inner fields), than a simple projection suffices: select those fields in the right order. The data then comply to `encoder.schema`.

Reply:
I'm OK to make the scope small and only support common cases like case class. But we must make sure the implementation is defensive and we give a clear better error message for unsupported cases.

Now this just throws attribute not found exception if the encoder is `Encoders.javaSerialization`.

Reply:
I have added checks to `toDS{T]` that throw a meaningful `AnalysisException` when columns do not line up with encoder's schema. The method is now general enough to support any encoder with a schema that can be created merely by projection, which is true for standard encoder of case classes (without classes that have inner case classes with extra fields or different field order), seq and maps of case classes (without extra fields or different field order) as well as any Java classes (`Encoders.javaSerialisation`). It throws `as[T]`'s Exception if columns do not line up with encoder's schema and `as[T]` would fail either.

Reply:
@cloud-fan what do you think about this implementation of `toDS[T]` and its scope?

Reply:
@cloud-fan any thoughts on this?

Reply:
@cloud-fan just pinging this thread to see if you have any more input here